README.md
&nbsp;

<div align="center">

<img src="https://raw.githubusercontent.com/AgnostiqHQ/covalent-awsbatch-plugin/main/assets/aws_batch_readme_banner.jpg" width=150%>

[![covalent](https://img.shields.io/badge/covalent-0.177.0-purple)](https://github.com/AgnostiqHQ/covalent)
[![python](https://img.shields.io/pypi/pyversions/covalent-awsbatch-plugin)](https://github.com/AgnostiqHQ/covalent-awsbatch-plugin)
[![tests](https://github.com/AgnostiqHQ/covalent-awsbatch-plugin/actions/workflows/tests.yml/badge.svg)](https://github.com/AgnostiqHQ/covalent-awsbatch-plugin/actions/workflows/tests.yml)
[![codecov](https://codecov.io/gh/AgnostiqHQ/covalent-awsbatch-plugin/branch/main/graph/badge.svg?token=QNTR18SR5H)](https://codecov.io/gh/AgnostiqHQ/covalent-awsbatch-plugin)
[![apache](https://img.shields.io/badge/License-Apache_License_2.0-blue)](https://www.apache.org/licenses/LICENSE-2.0)

</div>

## Covalent AWS Batch Plugin

Covalent is a Pythonic workflow tool used to execute tasks on advanced computing hardware.

This executor plugin interfaces Covalent with [AWS Batch](https://docs.aws.amazon.com/batch/) which allows tasks in a covalent workflow to be executed as AWS batch jobs.

## 1. Installation

To use this plugin with Covalent, simply install it using `pip`:

```
pip install covalent-awsbatch-plugin
```

## 2. Usage Example

This is an example of how a workflow can be adapted to utilize the AWS Batch Executor. Here we train a simple Support Vector Machine (SVM) model and use an existing AWS Batch Compute environment to run the `train_svm` electron as a batch job. We also note we require [DepsPip](https://covalent.readthedocs.io/en/latest/concepts/concepts.html#depspip) to install the dependencies when creating the batch job.

```python
from numpy.random import permutation
from sklearn import svm, datasets
import covalent as ct

deps_pip = ct.DepsPip(
	packages=["numpy==1.23.2", "scikit-learn==1.1.2"]
)

executor = ct.executor.AWSBatchExecutor(
    s3_bucket_name = "covalent-batch-qa-job-resources",
    batch_queue = "covalent-batch-qa-queue",
    batch_execution_role_name = "ecsTaskExecutionRole",
    batch_job_role_name = "covalent-batch-qa-job-role",
    batch_job_log_group_name = "covalent-batch-qa-log-group",
    vcpu = 2, # Number of vCPUs to allocate
    memory = 3.75, # Memory in GB to allocate
    time_limit = 300, # Time limit of job in seconds
)

# Use executor plugin to train our SVM model.
@ct.electron(
    executor=executor,
    deps_pip=deps_pip
)
def train_svm(data, C, gamma):
    X, y = data
    clf = svm.SVC(C=C, gamma=gamma)
    clf.fit(X[90:], y[90:])
    return clf

@ct.electron
def load_data():
    iris = datasets.load_iris()
    perm = permutation(iris.target.size)
    iris.data = iris.data[perm]
    iris.target = iris.target[perm]
    return iris.data, iris.target

@ct.electron
def score_svm(data, clf):
    X_test, y_test = data
    return clf.score(
    	X_test[:90],
	 	y_test[:90]
    )

@ct.lattice
def run_experiment(C=1.0, gamma=0.7):
    data = load_data()
    clf = train_svm(
    	data=data,
    	C=C,
    	gamma=gamma
    )
    score = score_svm(
    	data=data,
	 	clf=clf
    )
    return score

# Dispatch the workflow
dispatch_id = ct.dispatch(run_experiment)(
	C=1.0,
	gamma=0.7
)

# Wait for our result and get result value
result = ct.get_result(dispatch_id=dispatch_id, wait=True).result

print(result)
```

During the execution of the workflow one can navigate to the UI to see the status of the workflow, once completed however the above script should also output a value with the score of our model.

```
0.9777777777777777
```


## 3. Configuration

There are many configuration options that can be passed in to the class `ct.executor.AWSBatchExecutor` or by modifying the [covalent config file](https://covalent.readthedocs.io/en/latest/how_to/config/customization.html) under the section `[executors.awsbatch]`

For more information about all of the possible configuration values visit our [read the docs (RTD) guide](https://covalent.readthedocs.io/en/latest/api/executors/awsbatch.html) for this plugin.

## 4. Required AWS Resources

In order to run your workflows with covalent there are a few notable AWS resources that need to be provisioned first.

For more information regarding which cloud resources need to be provisioned visit our [read the docs (RTD) guide](https://covalent.readthedocs.io/en/latest/api/executors/awsbatch.html) for this plugin.

The required AWS resources include a Batch Job Definition, Batch Job Role, Batch Queue, Batch Compute Environment, Log Group, Subnet, VPC, and an S3 Bucket.

## Getting Started with Covalent


For more information on how to get started with Covalent, check out the project [homepage](https://github.com/AgnostiqHQ/covalent) and the official [documentation](https://covalent.readthedocs.io/en/latest/).

## Release Notes

Release notes for this plugin are available in the [Changelog](https://github.com/AgnostiqHQ/covalent-awsbatch-plugin/blob/main/CHANGELOG.md).

## Citation

Please use the following citation in any publications:

> W. J. Cunningham, S. K. Radha, F. Hasan, J. Kanem, S. W. Neagle, and S. Sanand.
> *Covalent.* Zenodo, 2022. https://doi.org/10.5281/zenodo.5903364

## License

Covalent is licensed under the Apache License 2.0. See the [LICENSE](https://github.com/AgnostiqHQ/covalent-executor-template/blob/main/LICENSE) file or contact the [support team](mailto:support@agnostiq.ai) for more details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [UNRELEASED]

## [0.42.0] - 2023-12-06

### Changed

- Edit terraform scripts to require no user input in most cases (maybe `region`).

## [0.41.2] - 2023-11-06

### Fixed

- Modified the `prefix` attribute from `ExecutorPluginDefaults` to take a default value, since it was causing error while using `deploy` command with infra arguments

## [0.41.1] - 2023-11-01

### Fixed

- Fixed config file generation from template by using newer `templatefile` for Mac M1 compatibility with terraform

## [0.41.0] - 2023-10-15

### Changed

- Added `container_image_uri` to `AWSBatchExecutor` to allow for custom container images to be specified via kwargs
- Corrected reference to past name in CONTRIBUTING.md
- Updated .env.example to remove outdated parameter

## [0.40.0] - 2023-10-13

### Changed

- Updated `requirements.txt`, removed boto3 version requirement as it is part of `covalent-aws-plugins`, and update `covalent-aws-plugins` minimum version requirement.

## [0.39.0] - 2023-10-06

### Changed

- Updated the terraform provider version for AWS to 5.17.0 (batch plugin)

## [0.38.0] - 2023-09-20

### Changed

- Updated license to Apache

## [0.37.0] - 2023-09-05

### Added

- Output stream and exceptions support for AWSBatchExecutor.

- A necessary `cache_dir` argument (in lieu of `**kwargs`, removed by previous).

## [0.36.0] - 2023-06-20

### Changed

- Updates __init__ signature kwargs replaced with parent for better documentation.

- Update task cancel function defintions. Add success, failure cancel tests

## [0.35.0] - 2023-05-02

### Changed

- Moved infra folder back under assets.

## [0.34.0] - 2023-05-02

### Changed

- In terraform `networking.tf`, `enable_nat_gateway` is set to 'false'.

## [0.33.0] - 2023-05-02

### Added

- Pydantic model validation for Batch executor class parameters.
- Pydantic model validation for the corresponding Terraform parameters.

### Changed

- Minor updates to the Terraform script such as changing variable `name` to `prefix`.

## [0.32.0] - 2023-03-14


### Added

- Adding `terraform` resource provisioning scripts to assets

## [0.31.0] - 2022-12-15

### Changed

- Removed references to `.env` file in the functional test README.

## [0.30.0] - 2022-12-14

### Changed

- Make Covalent Base Executor image configurable via environment variables.

## [0.29.0] - 2022-12-06

### Changed

- Using executor aliases instead of classes for functional tests

## [0.28.0] - 2022-12-06

### Changed

- Using region value directly from boto3 session to configure logging to support cases where user supplied region is empty

## [0.27.0] - 2022-11-28

### Changed

- Moved creation of temp directory outside constructor (client side) so it can be run on dispatcher side

## [0.26.0] - 2022-11-22

### Changed

- Removed hardcoded region, updated functional test executor fixture to include region, and added .env.example 

## [0.25.0] - 2022-11-22

### Changed

- Not setting default values for profile, region, and credentials_file

## [0.24.0] - 2022-11-15

### Changed

- Functional tests using pytest and .env file configuration

## [0.23.0] - 2022-10-28

### Changed

- Bumped aws plugins version to new stable release

## [0.22.0] - 2022-10-27

### Changed

- Added Alejandro to paul blart group

## [0.21.0] - 2022-10-27

### Fixed

- Changelog

### Changed

- pre-commit autoupdate
- Added license workflow

## [0.20.1] - 2022-10-27

### Fixed

- Fixed parallel execution of electrons submitting jobs to batch

## [0.20.0] - 2022-10-27

### Changed 

- Updated tag of hardcoded ECR URI to `stable`

## [0.19.0] - 2022-10-25

### Changed 

- Updated covalent-aws-plugins version `>=0.7.0rc0`.
- Cleanup file based method

## [0.18.0] - 2022-10-25

### Changed

- Pinned version of covalent-aws-plugins to 0.5.0rc0 

## [0.17.0] - 2022-10-18

### Changed

- Updated `boto3` calls to make them compatible with the async library.

### Docs

- Updated docs to include more information about required/optional config values, and provide information about each cloud resource that needs to be provisioned 

## [0.16.1] - 2022-10-04

### Fixed

- Store `BASE_COVALENT_AWS_PLUGINS_ONLY` in a temporary file rather than storing it as an environment variable.

## [0.16.0] - 2022-10-04

### Changed

- Setting `BASE_COVALENT_AWS_PLUGINS_ONLY` environment system wide to circumvent `setup.py` subprocesses when installing.

## [0.15.0] - 2022-09-30

### Added

-  Logic to specify that only the base covalent-aws-plugins package is to be installed.

## [0.14.1] - 2022-09-20

### Fixed

- Using `get_config` to get default configuration when init parameters are not supplied

## [0.14.0] - 2022-09-15

### Changed

- Updated requirements.txt to pin aws executor plugins to pre-release version 0.1.0rc0

### Fixed

- Added missing await in asyncio.sleep during polling

## [0.13.0] - 2022-09-15

### Changed


tests/test_utils.py
# Copyright 2021 Agnostiq Inc.
#
# This file is part of Covalent.
#
# Licensed under the Apache License 2.0 (the "License"). A copy of the
# License may be obtained with this software package or at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Use of this file is prohibited except in compliance with the License.
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Unit tests for AWS Batch executor utils file."""

import tempfile
from functools import partial
from pathlib import Path

import cloudpickle as pickle
import pytest

from covalent_awsbatch_plugin.utils import _execute_partial_in_threadpool, _load_pickle_file


@pytest.mark.asyncio
async def test_execute_partial_in_threadpool():
    """Test method to execute partial function in asyncio threadpool."""

    def test_func(x):
        return x

    partial_func = partial(test_func, x=1)
    future = await _execute_partial_in_threadpool(partial_func)
    assert future == 1


def test_load_pickle_file(mocker):
    """Test the method used to load the pickled file and delete the file afterwards."""
    temp_fp = "/tmp/test.pkl"
    with open(temp_fp, "wb") as f:
        pickle.dump("test success", f)

    assert Path(temp_fp).exists()
    res = _load_pickle_file(temp_fp)
    assert res == "test success"
    assert not Path(temp_fp).exists()


tests/functional_tests/svm_workflow.py
# Copyright 2021 Agnostiq Inc.
#
# This file is part of Covalent.
#
# Licensed under the GNU Affero General Public License 3.0 (the "License").
# A copy of the License may be obtained with this software package or at
#
#      https://www.gnu.org/licenses/agpl-3.0.en.html
#
# Use of this file is prohibited except in compliance with the License. Any
# modifications or derivative works of this file must retain this copyright
# notice, and modified files must contain a notice indicating that they have
# been altered from the originals.
#
# Covalent is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the License for more details.
#
# Relief from the License may be granted by purchasing a commercial license.


import sys

import covalent as ct
import executor_instance
from numpy.random import permutation
from sklearn import datasets, svm

deps_pip = ct.DepsPip(packages=["numpy==1.22.4", "scikit-learn==1.1.2"])


@ct.electron
def load_data():
    iris = datasets.load_iris()
    perm = permutation(iris.target.size)
    iris.data = iris.data[perm]
    iris.target = iris.target[perm]
    return iris.data, iris.target


@ct.electron(executor=executor_instance.executor, deps_pip=deps_pip)
def train_svm(data, C, gamma):
    X, y = data
    clf = svm.SVC(C=C, gamma=gamma)
    clf.fit(X[90:], y[90:])
    return clf


@ct.electron
def score_svm(data, clf):
    X_test, y_test = data
    return clf.score(X_test[:90], y_test[:90])


@ct.lattice
def run_experiment(C=1.0, gamma=0.7):
    data = load_data()
    clf = train_svm(data=data, C=C, gamma=gamma)
    score = score_svm(data=data, clf=clf)
    return score


dispatchable_func = ct.dispatch(run_experiment)

dispatch_id = dispatchable_func(C=1.0, gamma=0.7)
result = ct.get_result(dispatch_id=dispatch_id, wait=True)
status = str(result.status)

print(result)

if status == str(ct.status.FAILED):
    print("Basic Workflow failed to run.")
    sys.exit(1)


covalent_awsbatch_plugin/utils.py
# Copyright 2021 Agnostiq Inc.
#
# This file is part of Covalent.
#
# Licensed under the Apache License 2.0 (the "License"). A copy of the
# License may be obtained with this software package or at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Use of this file is prohibited except in compliance with the License.
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Helper methods for AWS Batch executor plugin."""

import asyncio
import json
import os

import cloudpickle as pickle


async def _execute_partial_in_threadpool(partial_func):
    loop = asyncio.get_running_loop()
    return await loop.run_in_executor(None, partial_func)


def _load_pickle_file(filename):
    """Method to load the pickle file."""
    with open(filename, "rb") as f:
        result = pickle.load(f)
    os.remove(filename)
    return result


def _load_json_file(filename):
    """Method to load the json file."""
    with open(filename, "r", encoding="utf-8") as f:
        result = json.load(f)
    os.remove(filename)
    return result


covalent_awsbatch_plugin/awsbatch.py
# Copyright 2021 Agnostiq Inc.
#
# This file is part of Covalent.
#
# Licensed under the Apache License 2.0 (the "License"). A copy of the
# License may be obtained with this software package or at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Use of this file is prohibited except in compliance with the License.
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""AWS Batch executor plugin for the Covalent dispatcher."""

import asyncio
from enum import Enum
import json
import os
import tempfile
from functools import partial
from pathlib import Path
from typing import Any, Callable, Dict, List, Optional, Tuple

import boto3
import botocore
import cloudpickle as pickle
from covalent.executor.schemas import ResourceMap, TaskSpec, TaskUpdate
from covalent._shared_files import TaskCancelledError, TaskRuntimeError
from covalent._shared_files.config import get_config
from covalent._shared_files.logger import app_log
from covalent._shared_files.util_classes import RESULT_STATUS, Status
from covalent_aws_plugins import AWSExecutor
from pydantic import BaseModel

from .utils import _execute_partial_in_threadpool, _load_json_file, _load_pickle_file


class ExecutorPluginDefaults(BaseModel):
    """
    Default configuration values for the executor
    """

    credentials: str = ""
    profile: str = ""
    region: str = "us-east-1"
    s3_bucket_name: str = "covalent-batch-job-resources"
    batch_queue: str = "covalent-batch-queue"
    batch_execution_role_name: str = "ecsTaskExecutionRole"
    batch_job_role_name: str = "covalent-batch-job-role"
    batch_job_log_group_name: str = "covalent-batch-job-logs"
    cache_dir: str = "/tmp/covalent"
    vcpu: int = 2
    memory: float = 3.75
    num_gpus: int = 0
    retry_attempts: int = 3
    time_limit: int = 300
    poll_freq: int = 10
    container_image_uri: str = "public.ecr.aws/covalent/covalent-executor-base:stable"


class ExecutorInfraDefaults(BaseModel):
    """
    Configuration values for provisioning AWS Batch cloud infrastructure
    """

    prefix: Optional[str] = "covalent-batch"
    aws_region: str = "us-east-1"
    vpc_id: str = ""
    subnet_id: str = ""
    instance_types: str = "optimal"
    min_vcpus: int = 0
    max_vcpus: int = 256
    aws_s3_bucket: str = "job-resources"
    aws_batch_queue: str = "queue"
    aws_batch_job_definition: str = "job-definition"
    credentials: Optional[str] = ""
    profile: Optional[str] = ""
    vcpus: Optional[int] = 2
    memory: Optional[float] = 3.75
    num_gpus: Optional[int] = 0
    retry_attempts: Optional[int] = 3
    time_limit: Optional[int] = 300
    poll_freq: Optional[int] = 10
    container_image_uri: Optional[str] = "public.ecr.aws/covalent/covalent-executor-base:stable"


_EXECUTOR_PLUGIN_DEFAULTS = ExecutorPluginDefaults().dict()

EXECUTOR_PLUGIN_NAME = "AWSBatchExecutor"

FUNC_FILENAME = "func-{dispatch_id}-{node_id}.pkl"
RESULT_FILENAME = "{dispatch_id}/{node_id}/result.tobj"
STDOUT_FILENAME = "{dispatch_id}/{node_id}/stdout.txt"
STDERR_FILENAME = "{dispatch_id}/{node_id}/stderr.txt"
QELECTRON_DB_FILENAME = "{dispatch_id}/{node_id}/qelectron_db.mdb"
SUMMARY_FILENAME = "{dispatch_id}/{node_id}/summary.json"

JOB_NAME = "covalent-batch-{dispatch_id}-{task_group_id}"


# Valid Batch job terminal statuses
class StatusEnum(str, Enum):
    CANCELLED = "CANCELLED"
    FAILED = "FAILED"
    SUCCEEDED = "SUCCEEDED"


class ReceiveModel(BaseModel):
    status: StatusEnum


class AWSBatchExecutor(AWSExecutor):
    # Opt into the new API
    SUPPORTS_MANAGED_EXECUTION = True

    """AWS Batch executor plugin class.

    Args:
        credentials: Full path to AWS credentials file.
        profile: Name of an AWS profile whose credentials are used.
        s3_bucket_name: Name of an S3 bucket where objects are stored.
        batch_queue: Name of the Batch queue used for job management.
        batch_execution_role_name: Name of the IAM role used by the Batch ECS agent.
        batch_job_role_name: Name of the IAM role used within the container.
        batch_job_log_group_name: Name of the CloudWatch log group where container logs are stored.
        vcpu: Number of vCPUs available to a task.
        memory: Memory (in GB) available to a task.
        num_gpus: Number of GPUs available to a task.
        retry_attempts: Number of times a job is retried if it fails.
        time_limit: Time limit (in seconds) after which jobs are killed.
        poll_freq: Frequency with which to poll a submitted task.
        cache_dir: Cache directory used by this executor for temporary files.
        container_image_uri: URI of the docker container image used by the executor.
    """

    def __init__(
        self,
        credentials: str = None,
        profile: str = None,
        region: str = None,
        s3_bucket_name: str = None,
        batch_queue: str = None,
        batch_execution_role_name: str = None,
        batch_job_role_name: str = None,
        batch_job_log_group_name: str = None,
        vcpu: int = None,
        memory: float = None,
        num_gpus: int = None,
        retry_attempts: int = None,
        time_limit: int = None,
        poll_freq: int = None,
        cache_dir: str = None,
        container_image_uri: str = None,
    ):
        super().__init__(
            region=region or get_config("executors.awsbatch.region"),
            credentials_file=credentials or get_config("executors.awsbatch.credentials"),
            profile=profile or get_config("executors.awsbatch.profile"),
            s3_bucket_name=s3_bucket_name or get_config("executors.awsbatch.s3_bucket_name"),
            execution_role=batch_execution_role_name
            or get_config("executors.awsbatch.batch_execution_role_name"),
            poll_freq=poll_freq or get_config("executors.awsbatch.poll_freq"),
            log_group_name=batch_job_log_group_name
            or get_config("executors.awsbatch.batch_job_log_group_name"),
        )

        self.batch_queue = batch_queue or get_config("executors.awsbatch.batch_queue")
        self.batch_job_role_name = batch_job_role_name or get_config(
            "executors.awsbatch.batch_job_role_name"
        )
        self.vcpu = vcpu or get_config("executors.awsbatch.vcpu")
        self.memory = memory or get_config("executors.awsbatch.memory")
        self.num_gpus = num_gpus or get_config("executors.awsbatch.num_gpus")
        self.retry_attempts = retry_attempts or get_config("executors.awsbatch.retry_attempts")
        self.time_limit = time_limit or get_config("executors.awsbatch.time_limit")
        self.container_image_uri = container_image_uri or get_config(
            "executors.awsbatch.container_image_uri"
        )

        self.cache_dir = cache_dir or get_config("executors.awsbatch.cache_dir")

        Path(self.cache_dir).mkdir(parents=True, exist_ok=True)

        config = {
            "profile": self.profile,
            "region": self.region,
            "credentials": self.credentials_file,
            "s3_bucket_name": self.s3_bucket_name,
            "batch_queue": self.batch_queue,
            "batch_job_role_name": self.batch_job_role_name,
            "batch_job_log_group_name": self.log_group_name,
            "batch_execution_role_name": self.execution_role,
            "vcpu": self.vcpu,
            "memory": self.memory,
            "num_gpus": self.num_gpus,
            "retry_attempts": self.retry_attempts,
            "time_limit": self.time_limit,
            "cache_dir": self.cache_dir,
            "poll_freq": self.poll_freq,
            "container_image_uri": self.container_image_uri,
        }

        self._debug_log("Starting AWS Batch Executor with config:")
        self._debug_log(config)

    def _debug_log(self, message):
        app_log.debug(f"AWS Batch Executor: {message}")

    async def run(self, function: Callable, args: List, kwargs: Dict, task_metadata: Dict):
        raise NotImplementedError

    async def submit_task(self):
        raise NotImplementedError

    async def _poll_task(self):
        raise NotImplementedError

    async def _upload_task(self):
        raise NotImplementedError

    async def query_result(self):
        raise NotImplementedError

    async def submit_task_group(
        self,
        task_specs: List[TaskSpec],
        resource_map: ResourceMap,
        task_group_metadata: Dict,
        identity: Dict
    ) -> Any:
        """
        Invokes the task on the remote backend.

        Args:
            task_metadata: Dictionary of metadata for the task. Current keys are `dispatch_id` and `node_id`.
            identity: Dictionary from _validate_credentials call { "Account": "AWS Account ID", ...}
        Return:
            task_uuid: Task UUID defined on the remote backend.
        """
        dispatch_id = task_group_metadata["dispatch_id"]
        task_group_id = task_group_metadata["task_group_id"]
        account = identity["Account"]

        self._debug_log("Submitting task...")
        boto_session = boto3.Session(**self.boto_session_options())
        batch = boto_session.client("batch")

        region = boto_session.region_name

        resources = [
            {"type": "VCPU", "value": str(int(self.vcpu))},
            {"type": "MEMORY", "value": str(int(self.memory * 1024))},
        ]
        if self.num_gpus:
            resources += [
                {
                    "type": "GPU",
                    "value": str(self.num_gpus),
                },
            ]

        COVALENT_TASK_SPECS = json.dumps([t.model_dump() for t in task_specs])
        COVALENT_RESOURCE_MAP = resource_map.model_dump_json()
        COVALENT_TASK_GROUP_METADATA = json.dumps(task_group_metadata)

        output_upload_uris = [
            (
                f"s3://{self.s3_bucket_name}/{RESULT_FILENAME.format(dispatch_id=dispatch_id, node_id=t.function_id)}",
                f"s3://{self.s3_bucket_name}/{STDOUT_FILENAME.format(dispatch_id=dispatch_id, node_id=t.function_id)}",
                f"s3://{self.s3_bucket_name}/{STDERR_FILENAME.format(dispatch_id=dispatch_id, node_id=t.function_id)}",
                f"s3://{self.s3_bucket_name}/{QELECTRON_DB_FILENAME.format(dispatch_id=dispatch_id, node_id=t.function_id)}",
            )
            for t in task_specs
        ]
        COVALENT_OUTPUT_UPLOAD_URIS = json.dumps(output_upload_uris)
        summary_upload_uris = [
            f"s3://{self.s3_bucket_name}/{SUMMARY_FILENAME.format(dispatch_id=dispatch_id, node_id=t.function_id)}"
            for t in task_specs
        ]
        COVALENT_SUMMARY_UPLOAD_URIS = json.dumps(summary_upload_uris)

        S3_STAGING_PREFIX = f"s3://{self.s3_bucket_name}/{dispatch_id}/sublattice_staging"

        # Register the job definition
        jobDefinitionName = f"{dispatch_id}-{task_group_id}"
        self._debug_log(f"Registering job definition {jobDefinitionName}...")
        env = [
            {
                "name": "COVALENT_TASK_SPECS",
                "value": COVALENT_TASK_SPECS,
            },
            {
                "name": "COVALENT_RESOURCE_MAP",
                "value": COVALENT_RESOURCE_MAP,
            },
            {
                "name": "COVALENT_TASK_GROUP_METADATA",
                "value": COVALENT_TASK_GROUP_METADATA,
            },
            {
                "name": "COVALENT_OUTPUT_UPLOAD_URIS",
                "value": COVALENT_OUTPUT_UPLOAD_URIS,
            },
            {
                "name": "COVALENT_SUMMARY_UPLOAD_URIS",
                "value": COVALENT_SUMMARY_UPLOAD_URIS,
            },
            {
                "name": "COVALENT_STAGING_URI_PREFIX",
                "value": S3_STAGING_PREFIX,
            },
        ]
        app_log.debug("Job Environment:")
        app_log.debug(env)
        partial_func = partial(
            batch.register_job_definition,
            jobDefinitionName=jobDefinitionName,
            type="container",  # Assumes a single EC2 instance will be used
            containerProperties={
                "environment": env,
                "image": self.container_image_uri,
                "jobRoleArn": f"arn:aws:iam::{account}:role/{self.batch_job_role_name}",
                "executionRoleArn": f"arn:aws:iam::{account}:role/{self.execution_role}",
                "resourceRequirements": resources,
                "logConfiguration": {
                    "logDriver": "awslogs",
                    "options": {
                        "awslogs-region": region,
                        "awslogs-group": self.log_group_name,
                        "awslogs-create-group": "true",
                        "awslogs-stream-prefix": "covalent-batch",
                    },
                },
            },
            retryStrategy={
                "attempts": self.retry_attempts,
            },
            timeout={
                "attemptDurationSeconds": self.time_limit,
            },
            platformCapabilities=["EC2"],
        )
        registered_job_definition = await _execute_partial_in_threadpool(partial_func)
        app_log.debug("Registered Job Definition:")
        app_log.debug(registered_job_definition)

        # Submit the job
        partial_func = partial(
            batch.submit_job,
            jobName=JOB_NAME.format(dispatch_id=dispatch_id, task_group_id=task_group_id),
            jobQueue=self.batch_queue,
            jobDefinition=jobDefinitionName,
        )
        response = await _execute_partial_in_threadpool(partial_func)
        return response["jobId"]

    async def get_status(self, job_id: str) -> Tuple[str, int]:
        """Query the status of a previously submitted Batch job.

        Args:
            batch: Batch client object.
            job_id: Identifier used to identify a Batch job.

        Returns:
            status: String describing the task status.
            exit_code: Exit code, if the task has completed, else -1.
        """
        self._debug_log("Checking job status...")
        batch = boto3.Session(**self.boto_session_options()).client("batch")
        partial_func = partial(batch.describe_jobs, jobs=[job_id])
        job = await _execute_partial_in_threadpool(partial_func)
        status = job["jobs"][0]["status"]

        self._debug_log(f"Got job status {status}")
        try:
            exit_code = int(job["jobs"][0]["container"]["exitCode"])
        except Exception as e:
            exit_code = -1
        return status, exit_code

    async def _poll_job(self, job_id: str) -> StatusEnum:
        """Poll a Batch job until completion."""
        self._debug_log(f"Polling Batch job with id {job_id}...")

        status, exit_code = await self.get_status(job_id)

        if status == "CANCELLED":
            return status

        while status not in ["SUCCEEDED", "FAILED"]:
            await asyncio.sleep(self.poll_freq)
            status, exit_code = await self.get_status(job_id)

        return status

    async def cancel(self, task_metadata: Dict, job_handle: str) -> bool:
        """
        Cancel the batch job.

        Args:
            task_metadata: Dictionary with the task's dispatch_id and node id.
            job_handle: Unique job handle assigned to the task by AWS Batch.

        Returns:
            If the job was cancelled or not
        """
        self._debug_log("Cancelling job ID {job_handle}...")
        try:
            batch = boto3.Session(**self.boto_session_options()).client("batch")
            partial_func = partial(
                batch.terminate_job,
                jobId=job_handle,
                reason=f"Triggered cancellation with {task_metadata}",
            )
            await _execute_partial_in_threadpool(partial_func)
            return True
        except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as error:
            app_log.debug(
                f"Failed to cancel AWS Batch job: {job_handle} with \
                          task_metadata: {task_metadata} with error:{error}"
            )
            return False

    async def send(
        self,
        task_specs: List[TaskSpec],
        resources: ResourceMap,
        task_group_metadata: Dict,
    ):

        dispatch_id = task_group_metadata["dispatch_id"]
        task_group_id = task_group_metadata["task_group_id"]

        batch_job_name = JOB_NAME.format(dispatch_id=dispatch_id, task_group_id=task_group_id)

        self._debug_log(f"Executing Dispatch ID {dispatch_id} Task Group {task_group_id}")

        self._debug_log("Validating Credentials...")
        partial_func = partial(self._validate_credentials, raise_exception=True)
        identity = await _execute_partial_in_threadpool(partial_func)

        if await self.get_cancel_requested():
            raise TaskCancelledError(f"AWS Batch job {batch_job_name} requested to be cancelled")

        job_id = await self.submit_task_group(task_specs, resources, task_group_metadata, identity)
        self._debug_log(f"Successfully submitted job with ID: {job_id}")
        await self.set_job_handle(handle=job_id)

        return job_id

    async def poll(self, task_group_metadata: Dict, job_id: str) -> Dict:
        return {"status": await self._poll_job(job_id)}

    async def receive(self, task_group_metadata: Dict, data: Any) -> List[TaskUpdate]:
        # Job should have reached a terminal state by the time this is invoked.
        dispatch_id = task_group_metadata["dispatch_id"]
        task_group_id = task_group_metadata["task_group_id"]

        received = ReceiveModel.model_validate(data)
        s3 = boto3.Session(**self.boto_session_options()).client("s3")
        task_updates = []

        # Try to retrieve whatever task artifacts have been uploaded
        # Electrons with a complete execution summary are either COMPLETED or FAILED.
        # Electrons with no execution summary are either FAILED or CANCELLED depending on the overall Batch job status

        for i, node_id in enumerate(task_group_metadata["node_ids"]):
            summary_filename = SUMMARY_FILENAME.format(dispatch_id=dispatch_id, node_id=node_id)

            try:
                resp = s3.get_object(Bucket=self.s3_bucket_name, Key=summary_filename)
                result_summary = json.loads(resp["Body"].read())
                status = RESULT_STATUS.FAILED if result_summary["exception_occurred"] else RESULT_STATUS.COMPLETED
                assets = {
                    "output": {
                        "remote_uri": result_summary["output_uri"],
                    },
                    "stdout": {
                        "remote_uri": result_summary["stdout_uri"],
                    },
                    "stderr": {
                        "remote_uri": result_summary["stderr_uri"],
                    },
                    "qelectron_db": {
                        "remote_uri": result_summary["qelectron_db_uri"],
                    },
                }
                update = {
                    "dispatch_id": dispatch_id,
                    "node_id": node_id,
                    "status": status,
                    "assets": assets,
                }
                task_updates.append(TaskUpdate(**update))

            except s3.exceptions.NoSuchKey:
                for j in range(i, len(task_group_metadata["node_ids"])):
                    status = RESULT_STATUS.CANCELLED if received.status == StatusEnum.CANCELLED else RESULT_STATUS.FAILED
                    update = {
                        "dispatch_id": dispatch_id,
                        "node_id": task_group_metadata["node_ids"][j],
                        "status": status,
                        "assets": {}
                    }
                    task_updates.append(TaskUpdate(**update))
                    break

        return task_updates

    def get_upload_uri(self, task_group_metadata: Dict, object_key: str) -> str:
        dispatch_id = task_group_metadata["dispatch_id"]
        gid = task_group_metadata["task_group_id"]
        s3_object_key = f"{dispatch_id}/uploads/{gid}/{object_key}"
        s3_bucket_name = self.s3_bucket_name
        return f"s3://{s3_bucket_name}/{s3_object_key}"
